### PR TITLE
Lightning enchantment

### PR DESCRIPTION
**Pull request by FordFriedel**
Adds "lightning" enchantment to swords, which strikes lightning and deals additional damage to entities when the player is below the sky during a storm.

### CODE REVIEW

**Review by oddlama — CHANGES_REQUESTED** [review]
Generally your code looks great! There are some minor formatting issues, but that's mostly me and my "religious beliefs" regarding that matter. Nothing an auto formatter wouldn't fix anyways.

Now that I see your enchantment I must say that I do like it. Although it arguably doesn't fit into the vane spirit regarding strictly useful additions (and there are already those trident thingies that have a similar feature). As the enchantment is really low-maintenance, I'd be willing to include your changes under the compromise that would be disabled by default. All the players who use vane for the "vanilla enhancements" stuff wouldn't be put off by a fun addition.

**EDIT:** If I'm not mistaken, you can add this function to your enchantment to disable it by default:
```
public boolean config_enabled_def() {
    return config_enabled_def;
}
```

File: vane-enchantments/src/main/java/org/oddlama/vane/enchantments/enchantments/Lightning.java
Comment (FordFriedel) [reply vane-enchantments/src/main/java/org/oddlama/vane/enchantments/enchantments/Lightning.java:88]: Agreed.

**Comment by FordFriedel** [comment]
So by writing the code: 

`public boolean config_enabled_def() {
    return false;
}` 

this will automatically write to the config file that our feature should be disabled by default? Or is further editing needed to achieve this.

**Comment by oddlama** [comment]
Yep, but it won't change existing configurations (so you might want to just delete yours to test it. It will be regenerated.)

You might have noticed that vane has it's own configuration system which makes it so easy to add configuration options anywhere in the code. That specific function here is a part of this special configuration system. It isn't an overridden function but behaves similar to one. If this function exists, the system will simply use it to determine the default enabled status for the submodule (the enchantment).

File: vane-enchantments/src/main/java/org/oddlama/vane/enchantments/enchantments/Lightning.java
Comment (FordFriedel) [reply vane-enchantments/src/main/java/org/oddlama/vane/enchantments/enchantments/Lightning.java:84]:
```java
    @ConfigInt(
        def = 4,
        min = 0,
        max = 20,
        desc = "Damage modifier for the lightning enchant"
    )
    private int config_lightning_damage;
```
Does this look correct?

Comment (oddlama) [reply vane-enchantments/src/main/java/org/oddlama/vane/enchantments/enchantments/Lightning.java]: looks good :)

**Review by oddlama — COMMENTED** [review]
Looks good so far! Next step would be testing that everything works as expected, and then it's finished I guess :)

**Comment by FordFriedel** [comment]
One problem before in game testing: the enchantment is enabled by default, do I need to `return true` to disable the enchantment by default?

**Comment by FordFriedel** [comment]
after a quick test its clear that I will want to protect the user from getting struck by their own lightning, maybe I take it one step further and make anyone yielding a lightning sword immune to taking damage from lightning

**Comment by FordFriedel** [comment]
Should be ready for re-review and formatting requests

File: vane-enchantments/src/main/java/org/oddlama/vane/enchantments/enchantments/Lightning.java
Comment (oddlama) [inline vane-enchantments/src/main/java/org/oddlama/vane/enchantments/enchantments/Lightning.java:39]: My bad, it seems like I didn't remember this correctly. I'm sorry. I've checked and this only works for regular "modules", i.e. classes that extend `ModuleGroup`. The fix is simple, but let me explain what modules are exactly, as it might be good to understand this for the future (and future readers).

All parts and features of vane are grouped into separate modules. Modules can have config options and language strings associated to them `@Config....`, and in theory each module can be disabled independently of one another. Sometimes we expose this functionality as an additional `enabled` config in that module.

To make it easier to enable related modules, they are put in a hierarchy. This means each module has a context to which it belongs, like a "parent module", but without directly imposing nesting which must be specified explicitly.

Each vane plugin (vane-enchantments) is a module in and of itself. Say the plugin as a config option `my_plugin_option` and we write two features as modules which have the config options `feature_option_a` and `feature_option_b`. When modules are created, they specify the context they belong to in their constructor like `super(context)`. So even if the modules of those features have the plugin as the context, the options will be on the _same level_ in the yaml file, as they belong to the same context:

```yaml
my_plugin_option: "test1"
feature_option_a: "test2"
feature_option_b: "test3"
```

If we want nesting, we create a new namespace subcontext for each feature module by calling `super(context.namespace("feature_a"))`. Then we get:

```yaml
my_plugin_option: "test1"
feature_a:
  feature_option_a: "test2"
feature_b:
  feature_option_b: "test3"
```

A common use case is to have a namespace contexts, which users should be able to enable or disable. That is called a group. The main plugin is always a group, as it has an implicit `enabled` field.

So each module is only active in vane, if its context is enabled. Enabled-ness propagates downwards in the context hierarchy. So whatever the context of a module is exactly, it only counts as enabled, if all parent groups in it's hierarchy are enabled. This makes it so easy to enable or disable related vane features without writing any additional code each time.

In the code we would create a group with `super(context.group("Group name")`, which results in this:

```yaml
my_plugin_option: "test1"
feature_a:
  enabled: true
  feature_option_a: "test2"
feature_b:
  enabled: true
  feature_option_b: "test3"
```

---

These groups basically registers it's own `enabled` config field, and if we want to change the default,
it must be done in that specific group. One can call `context.group("name", default_enabled)` for convenience already. So the merit is that this it isn't part of the class that calls `super(context.....)`, which in our case is the enchantment. And that's why the override didn't work (The enabled config is not directly a part of the enchantment's context).

So we need to set this property when creating the context group for the enchantment. To make this easier, I've pushed a new feature that adds a constructor parameter to specify the default enabled value. This means you should rebase upon the newest commit and remove this function here entirely and instead do this:

```
public Lightning(Context<Enchantments> context) {
    // Don't enable this enchantment by default
    super(context, false);
}
```

I hope that explains things a little. If not please ask me to clarify!

**Review by oddlama — CHANGES_REQUESTED** [review]

**Comment by oddlama** [comment]
Looks very very good! Please see the comment I wrote, there's just one little thing that needs fixing because I told you something incorrect (sorry). After that we can merge :)

**Comment by FordFriedel** [comment]
Sorry I am just a little confused. I understand the idea of modules in vane to allow for users to customize their experience by toggling features on and off.  

What I dont understand is do you want me to return true? I believe I had that previously and that turned it on by default.

**Comment by oddlama** [comment]
What I want you to do is to remove the function `config_enabled_def` completely and instead write add this constructor:

```
public Lightning(Context<Enchantments> context) {
    // Don't enable this enchantment by default
    super(context, false);
}
```

**Comment by oddlama** [comment]
Yes, that's why I stated you need to rebase on the newest commit on the develop branch. I just added this parameter yesterday for you :)

**Comment by oddlama** [comment]
If you don't want the hassle with rebasing and switching to 1.19 i can also do this myself if need be. It's important that you rebase and not merge to preserve a clean history.

**Comment by FordFriedel** [comment]
After a stressful bit I have rebased.  **I believe** that it's ready to go.  Thanks.

**Review by oddlama — APPROVED** [review]
Thanks for your hard work! Sorry for causing you to go through the rebasing trouble, but I guess this also is an essential part of writing PRs :P. I'll become second nature.

I also agree that this looks ready to merge now. I'm a bit busy at the moment, but will merge this when I find time. I'll also release a new minor version as soon as paper 1.19 gets out of experimental, just in case I have to fix some other things which might pop up over the next days.

**Comment by FordFriedel** [comment]
Sounds good! I am trying to accrue more topics to my repertoire so I appreciate having to struggle through new things.  Big thank you for giving me a platform to learn through!

**Pull request merged**
